File: in_Java/0213 House Robber II.java
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        int[] a1 = new int[nums.length-1];
        int[] a2 = new int[nums.length-1];
        for(int i = 0; i < a1.length; i++){
            a1[i] = nums[i];
            a2[i] = nums[i+1];
        }
        int[] dp1 = new int[a1.length+1];
        int[] dp2 = new int[a2.length+1];
        dp1[1] = a1[0];
        dp2[1] = a2[0];
        for(int i = 2; i < dp1.length; i++){
            dp1[i] = Math.max(dp1[i-1], dp1[i-2] + a1[i-1]);
            dp2[i] = Math.max(dp2[i-1], dp2[i-2] + a2[i-1]);
        }
        return Math.max(dp1[dp1.length-1], dp2[dp2.length-1]);
    }
}

File: in_Python/0213 House Robber II.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        def helper(arr):
            if arr == None or len(arr) == 0:
                return 0
            N = len(arr)
            dp = [0] * (N+1)
            dp[1] = arr[0]
            for i in range(2, len(dp)):
                dp[i] = max(dp[i-1], dp[i-2] + arr[i-1])
            return dp[-1]
        
        n = len(nums)
        if n == 1:
            return nums[0]
        arr1 = nums[0:n-1]
        arr2 = nums[1:]
        x1 = helper(arr1)
        x2 = helper(arr2)
        return max(x1,x2)